src/EventTeachingService.cpp

```diff
@@ -11,259 +11,427 @@
+      }      // for me
 
-  case OPC_NERD:
-    // request for all stored events
-    // DEBUG_SERIAL << F("> NERD : request all stored events for nn = ") << nn << endl;
+      return PROCESSED;
 
-    if (nn == module_config->nodeNum)
-    {
-      msg->len = 8;
-      msg->data[0] = OPC_ENRSP;                       // response opcode
-      msg->data[1] = highByte(module_config->nodeNum);        // my NN hi
-      msg->data[2] = lowByte(module_config->nodeNum);         // my NN lo
+    case OPC_NENRD:
+      // 72 - request read stored event by event index
+      // DEBUG_SERIAL << F("> NERD : request all stored events for nn = ") << nn << endl;
 
-      for (byte i = 0; i < module_config->EE_MAX_EVENTS; i++)
+      if (nn == module_config->nodeNum) 
       {
-        if (module_config->getEvTableEntry(i) != 0) {
-          // it's a valid stored event
+        msg->len = 8;
+        msg->data[0] = OPC_ENRSP;     // response opcode
+        msg->data[1] = highByte(nn);  // my NN hi
+        msg->data[2] = lowByte(nn);   // my NN lo
 
+        byte i = msg->data[3];
+        if ((i >= module_config->EE_MAX_EVENTS) && (module_config->getEvTableEntry(i) == 0)) 
+        {
+          controller->sendGRSP(OPC_REVAL, getServiceID(), CMDERR_INV_EN_IDX);
+          controller->sendCMDERR(CMDERR_INV_EN_IDX);
+        } 
+        else 
+        {
+          // it's a valid stored event
           // read the event data from EEPROM
           // construct and send a ENRSP message
           module_config->readEvent(i, &msg->data[3]);
-          msg->data[7] = i;                           // event table index
+          msg->data[7] = i;  // event table index
 
           // DEBUG_SERIAL << F("> sending ENRSP reply for event index = ") << i << endl;
           controller->sendMessage(msg);
-          delay(10);
-
-        } // valid stored ev
-      } // loop each ev
-    } // for me
+        }
+      }  // for me
 
-    return PROCESSED;
+      return PROCESSED;
 
-  case OPC_REVAL:
-    // received REVAL -- request read of an event variable by event index and ev num
-    // respond with NEVAL
+    case OPC_REVAL:
+      // 9C - request read of an event variable by event index and ev num
+      // respond with NEVAL
 
-    if (nn == module_config->nodeNum)
-    {
-      if (module_config->getEvTableEntry(msg->data[3]) != 0)
-      {
-        byte value = module_config->getEventEVval(msg->data[3], msg->data[4]);
-        controller->sendMessageWithNN(OPC_NEVAL, msg->data[3], msg->data[4], value);
-      }
-      else
+      if (nn == module_config->nodeNum) 
       {
-        // DEBUG_SERIAL << F("> request for invalid event index") << endl;
-        controller->sendCMDERR(6);
+        if (msg->len < 5) 
+        {
+          controller->sendGRSP(OPC_REVAL, getServiceID(), CMDERR_INV_CMD);
+          controller->sendCMDERR(CMDERR_INV_CMD);
+        } 
+        else 
+        {
+          if (module_config->getEvTableEntry(msg->data[3]) != 0) 
```

Comment:
Should we also check if the event index is too large?

Reply:
This is checked in configuration.cpp which returns 0 if the index is larger than EE_MAX_EVENTS.  So, if it is too large it runs the else code and sends error messages.